waspc/src/Analyzer/Parser/Util.hs

```diff
@@ -0,0 +1,66 @@
+module Analyzer.Parser.Util
+  ( ParserState (..),
+    initialState,
+    Parser,
+    updatePosition,
+    putInput,
+    setStartCode,
+    ParserInput,
+  )
+where
+
+import Analyzer.Parser.ParseError (ParseError)
+import Analyzer.Parser.Token (SourcePosition (..))
+import Control.Monad.Trans.Except (Except)
+import Control.Monad.Trans.State.Lazy (StateT, get, modify)
+import Data.Word (Word8)
+
+-- | Tracks state of the parser, which is
+--   - Current line/column position
+--   - Current input to the lexer
+data ParserState = ParserState
+  { parserSourcePosition :: SourcePosition,
+    parserRemainingInput :: ParserInput,
+    parserStartCode :: Int
```

Comment:
This can be confusing for somebody who doesn't know what start code is. I wonder if we should document this in some way, or just let people figure it out on their own. Hmmmm.

Reply:
Start codes exist both in lexer and parser, right? In this case, it says parserStartCode, but it is actually lexer start code. It is not very wrong to call it parserStartCode since lexer is part of the parser, but hm, we are actually specifically using this field as lexer start code. So calling it `parserLexerStartCode` might be the best, since it is most precise, and we can just provide link to Alex docs next to it.